hyperspy/io_plugins/nexus.py

```diff
@@ -0,0 +1,1130 @@
+    axis_names=[_parse_to_file(".")]*len(signal.axes_manager.shape)
+    for i,axis in enumerate(signal.axes_manager._axes):
+        if axis.name != t.Undefined:
+            axname = _parse_to_file(axis.name)
+            axindex = [axis.index_in_array]
+            indices = _parse_to_file(axis.name + "_indices")
+            nxdata.attrs[indices] = _parse_to_file(axindex)
+            #navigate = _parse_to_file(axis.name + "_navigation")
+            #nxdata.attrs[navigate] = _parse_to_file(axis.navigate)
+            dset = nxdata.require_dataset(axname,data=axis.axis,shape=axis.axis.shape,
+                                   dtype=axis.axis.dtype)
+            dset.attrs["navigation"] = _parse_to_file(axis.navigate)
+            axis_names[axis.index_in_array]=axname
+
+    nxdata.attrs["axes"] = _parse_to_file(axis_names)
+    return nxdata
+
+def file_writer(filename,
+                signals,
+                save_original_metadata=True,
+                *args, **kwds):
+    """ Write the signal and metadata as a nexus file
+    This will save the signal in NXdata format in the file.
+    As the form of the metadata can vary and is not validated it will 
+    be stored as an NXcollection (an unvalidated collection)
+
+    Parameters
+    ----------
+    filename : str    
+        Path of the file to write
+    signals : signal or list of signals
+    store_original_metadata : bool , default : False
+          Option to store hyperspy.original_metadata with the signal
+          A loaded Nexus file may have a large amount of data
+          when loaded which you may wish to omit on saving
+          
+    See Also 
+    --------
+    * :py:meth:`~.io_plugins.nexus.file_reader`
+    * :py:meth:`~.io_plugins.nexus.list_datasets_in_file`
+    * :py:meth:`~.io_plugins.nexus.read_metadata_from_file`
+         
+    """
+    if not isinstance(signals,list):
+        signals = [signals]
+
+    with h5py.File(filename, mode='w') as f:
+        f.attrs['file_format'] = "nexus_like"
+        f.attrs['file_writer'] = "hyperspy_nexus"        
+        nxaux = f.create_group("auxiliary")
+        nxaux.attrs["NX_class"] = _parse_to_file("NXentry")                
+        nxexpt = f.create_group('signals')
+        nxexpt.attrs["NX_class"] = _parse_to_file("NXentry")               
+        if 'compression' not in kwds:
+            kwds['compression'] = 'gzip'
+
+        #
+        # write the signals
+        #
+
+        for i,sig in enumerate(signals):
+            signal_name = sig.metadata.General.title if \
+            sig.metadata.General.title else 'unnamed__%d'% i
+            if "/" in signal_name:
+                signal_name = signal_name.replace("/", "_")
+            if signal_name.startswith("__"):
+                signal_name = signal_name[2:]
+            # store type of signal
+            if hasattr(sig,"_signal_type"):
+                sig.metadata.Signal.signal_type = sig._signal_type
```

Comment:
Ok will remove - I wanted to store the signal_type but the metadata signal_type set isn't set when creating a signal1d or signal2d  e.g. via hs.signals   or as_signal1D.  
signal_type only seems to be set for a signal outside the core signals - e.g. EDSSEMSpectrum 
so if I remove it will only store "" for signal 1D and 2D.

Reply:
Indeed, this is expect behaviour: `Signal1D` and `Signal2D` are generic and therefore their `signal_type` is not defined. When creating/changing signal class, this is a logic which looks at signal dimension to assign the correct class:
https://github.com/hyperspy/hyperspy/blob/2229ffde051af21d3adc7ae626a72bf769c84374/hyperspy/io.py#L404-L479